### PR TITLE
[UI] As an Administrator, I want to select period of graphs and counts on the dashboard (-Day/2 Weeks/Month/Year)

### PR DESCRIPTION
**Pull request by SirDomin**
| Q               | A
| --------------- | -----
| Branch?         | master
| Bug fix?        | no
| New feature?    | yes
| BC breaks?      | not sure
| Deprecations?   | not sure
| Related tickets | 
| License         | MIT

day not supported yet as a period

image

### CODE REVIEW

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 1]
```diff
@@ -0,0 +1,85 @@
+<?php
+
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:2]: missing DocBlock

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+{
+    /** @var EngineInterface */
+    private $templatingEngine;
+
+    /** @var DashboardStatisticsProviderInterface */
+    private $statisticsProvider;
+
+    /**
+     * @var SalesDataProviderInterface
+     */
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:25]:
```suggestion
    /** @var SalesDataProviderInterface */
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        $salesSummary = $this->salesDataProvider->getSalesSummary(
+            $channel,
+            $startDate,
+            $endDate,
+            Interval::{$interval}()
+        );
+
+        return [
+            'sales_summary' => [
+                'months'=>$salesSummary->getIntervals(),
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:62]:
```suggestion
                'months' => $salesSummary->getIntervals(),
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+                'sales' => $salesSummary->getSales()
+            ],
+            'channel'=>[
+              'base_currency_code'=>$channel->getBaseCurrency()->getCode()
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:66]:
```suggestion
              'base_currency_code' => $channel->getBaseCurrency()->getCode()
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+            $endDate,
+            Interval::{$interval}()
+        );
+
+        return [
+            'sales_summary' => [
+                'months'=>$salesSummary->getIntervals(),
+                'sales' => $salesSummary->getSales()
+            ],
+            'channel'=>[
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:65]:
```suggestion
            'channel'=>[
```
```suggestion
            'channel' => [
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 6]
```diff
@@ -0,0 +1,85 @@
...
+            'statistics'=>[
+                'total_sales' => $statistics->getTotalSales(),
+                'number_of_new_orders' => $statistics->getNumberOfNewOrders(),
+                'number_of_new_customers' => $statistics->getNumberOfNewCustomers(),
+                'average_order_value' => $statistics->getAverageOrderValue(),
+            ]
+        ];
+    }
+
+    public function getJsonData(ChannelInterface $channel, $interval): Response
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:77]: wrong order of functions, should be `public` => `protected` => `private`

**Comment by AdamKasp** [comment]
screens or clip with new graph will be amazing :)

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 7]
```diff
@@ -0,0 +1,85 @@
...
+    {
+        return $this->templatingEngine->renderResponse(
+            '@SyliusAdmin/Dashboard/Statistics/_template.html.twig', $this->getRawData($channel,
+            (new \DateTime('first day of next month last year')),
+            (new \DateTime('last day of this month')),
+            'month')
+        );
+    }
+
+    private function getRawData(ChannelInterface $channel, $startDate, $endDate, $interval): array
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:48]: I believe it could be the only method in this service (and, of course, public). It seems strange (and too complicated), to have a service that is used both as a controller for rendering templates and that provides data for other services. Let's simplify it - name this service `DashboardDataProvider` with one method `getRawData` returning `array` or some VO and leave the twig- and request/response-related work for the `DashboardController`. WDYT? ⚖️

File: src/Sylius/Bundle/AdminBundle/Resources/config/routing.yml [hunk 8]
```diff
@@ -11,6 +11,11 @@ sylius_admin_dashboard:
     defaults:
         _controller: sylius.controller.admin.dashboard:indexAction
 
+sylius_admin_dashboard_statistics:
+    path: /stats
```
Comment (Zales0123) [inline src/Sylius/Bundle/AdminBundle/Resources/config/routing.yml:15]: `/statistics` (abbreviations are evil 😈)

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 9]
```diff
@@ -0,0 +1,101 @@
...
+        $this->moneyFormatter = $moneyFormatter;
+    }
+
+    public function __invoke(ChannelInterface $channel): Response
+    {
+        return $this->templatingEngine->renderResponse(
+            '@SyliusAdmin/Dashboard/Statistics/_template.html.twig', $this->getRawData($channel,
+            (new \DateTime('first day of january this year')),
+            (new \DateTime('first day of january next year')),
+            'month')
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:57]: I assume, we've resigned from the option of skipping data in the twig load?
```suggestion
            '@SyliusAdmin/Dashboard/Statistics/_template.html.twig', 
            $this->getRawData(
                $channel,
                new \DateTime('first day of january this year'),
                new \DateTime('first day of january next year'),
                'month'
            )
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+        );
+    }
+
+    public function getJsonData(ChannelInterface $channel, $interval, $startDate, $endDate): Response
+    {
+        return new JsonResponse($this->getRawData($channel,
+            (new \DateTime($startDate)),
+            (new \DateTime($endDate)),
+            $interval)
+        );
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:67]:
```suggestion
        return new JsonResponse($this->getRawData(
            $channel,
            new \DateTime($startDate),
            new \DateTime($endDate),
            $interval
        ));
```
We used to align all arguments with the same indent.

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 11]
```diff
@@ -38,72 +33,58 @@ final class DashboardController
...
+    {
+        /** @var ChannelInterface|null $channel */
+        $channel = $this->findChannelByCodeOrFindFirst($request->query->get('channel'));
+
+        if (null === $channel) {
+            return new RedirectResponse($this->router->generate('sylius_admin_channel_create'));
         }
 
-        return $this->templatingEngine->renderResponse('@SyliusAdmin/Dashboard/index.html.twig', $data);
+        return $this->statisticsController->getJsonData(
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:74]: Triggering of a controller in another controller is always a bad idea. The common logic needs to be extracted to separate service.

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 12]
```diff
@@ -38,72 +33,58 @@ final class DashboardController
...
         EngineInterface $templatingEngine,
         RouterInterface $router,
-        ?SalesDataProviderInterface $salesDataProvider = null
+        ?StatisticsController $statisticsController
     ) {
-        $this->statisticsProvider = $statisticsProvider;
         $this->channelRepository = $channelRepository;
         $this->templatingEngine = $templatingEngine;
         $this->router = $router;
-        $this->salesDataProvider = $salesDataProvider;
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:55]: No one simply removed a parameter from the constructor once, it was in the tagged version. Both services need to be brought back due to BC.

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 13]
```diff
@@ -0,0 +1,101 @@
...
+        SalesDataProviderInterface $salesDataProvider,
+        MoneyFormatterInterface $moneyFormatter
+    ) {
+        $this->templatingEngine = $templatingEngine;
+        $this->statisticsProvider = $statisticsProvider;
+        $this->salesDataProvider = $salesDataProvider;
+        $this->moneyFormatter = $moneyFormatter;
+    }
+
+    public function __invoke(ChannelInterface $channel): Response
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:51]: The rule of thumb for me would be, that if we have more than one function in the class, none of them should be called `__invoke`. 
```suggestion
    public function renderStatistics(ChannelInterface $channel): Response
```

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 14]
```diff
@@ -38,72 +33,58 @@ final class DashboardController
...
-                ->getSalesSummary($channel, $startDate, $endDate, $interval)
-            ;
-            $data['currency'] = $channel->getBaseCurrency()->getCode();
+
+        return $this->templatingEngine->renderResponse('@SyliusAdmin/Dashboard/index.html.twig', [
+            'channel' => $channel,
+        ]);
+    }
+
+    public function getRawData(Request $request): Response
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:65]: Shouldn't this be part of StatisticController? Is it really the responsibility of this class to render this data as well?

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php (hunk 12 above)
Comment (SirDomin) [reply src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:55]: i got it back !

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 15]
```diff
@@ -0,0 +1,82 @@
...
+{
+    function let(DashboardStatisticsProviderInterface $statisticsProvider, SalesDataProviderInterface $salesDataProvider, MoneyFormatterInterface $moneyFormatter): void
+    {
+        $this->beConstructedWith($statisticsProvider, $salesDataProvider, $moneyFormatter);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(StatisticsDataProvider::class);
+    }
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:27]: you can omit this test but you should add test to check if this class implements the proper interface 
```suggestion
    function it_implements_a_statistics_data_provider_interface(): void
    {
        $this->shouldImplement(StatisticsDataProviderInterface::class);
    }
```

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 16]
```diff
@@ -0,0 +1,82 @@
...
+        SalesSummaryInterface $salesSummary,
+        CurrencyInterface $currency,
+        MoneyFormatterInterface $moneyFormatter
+    ): void {
+        $startDate = new \DateTime('2020-01-01');
+        $endDate = new \DateTime('2021-01-01');
+        $interval = 'month';
+        $currencyCode = 'USD';
+        $totalSales = 37274;
+        $averageOrderValue= 18637;
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:43]: imo unnecessary variables.

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 17]
```diff
@@ -0,0 +1,82 @@
...
+    {
+        $this->beConstructedWith($statisticsProvider, $salesDataProvider, $moneyFormatter);
+    }
+
+    function it_is_initializable()
+    {
+        $this->shouldHaveType(StatisticsDataProvider::class);
+    }
+
+    function it_provides_data(
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:29]: missing test for different intervals and periods

File: src/Sylius/Component/Core/Dashboard/DashboardStatistics.php [hunk 18]
```diff
@@ -24,14 +27,23 @@ class DashboardStatistics
...
-    public function __construct(int $totalSales, int $numberOfNewOrders, int $numberOfNewCustomers)
+    public function __construct(int $totalSales, int $numberOfNewOrders, int $numberOfNewCustomers, ChannelInterface $channel = null)
     {
         $this->totalSales = $totalSales;
         $this->numberOfNewOrders = $numberOfNewOrders;
         $this->numberOfNewCustomers = $numberOfNewCustomers;
+        $this->channel = $channel;
+    }
+
+    public function getChannel(): ?ChannelInterface
```
Comment (AdamKasp) [inline src/Sylius/Component/Core/Dashboard/DashboardStatistics.php:44]: you should update the spec too

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+
+    /** @var StatisticsDataProviderInterface */
+    private $statisticsDataProvider;
+
+    public function __construct(
+        EngineInterface $templatingEngine,
+        DashboardStatisticsProviderInterface $statisticsProvider,
+        SalesDataProviderInterface $salesDataProvider,
+        MoneyFormatterInterface $moneyFormatter,
+        StatisticsDataProviderInterface $statisticsDataProvider
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php:47]: Do you need all these services? You only use `templatingEngine` and `statisticsDataProvider` in the method below

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 20]
```diff
@@ -0,0 +1,66 @@
...
+            $channel,
+            $startDate,
+            $endDate,
+            Interval::{$interval}()
+        );
+
+        /** @var string */
+        $currencyCode = $channel->getBaseCurrency()->getCode();
+
+        return [
```
Comment (AdamKasp) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:49]: `getRawData` function returns a really scary array with 2 levels, WDYT about using here some VO?

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 21]
```diff
@@ -0,0 +1,66 @@
...
+    /** @var DashboardStatisticsProviderInterface */
+    private $statisticsProvider;
+
+    /** @var SalesDataProviderInterface */
+    private $salesDataProvider;
+
+    /** @var MoneyFormatterInterface */
+    private $moneyFormatter;
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:22]:
```suggestion
```

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 22]
```diff
@@ -0,0 +1,66 @@
...
+        DashboardStatisticsProviderInterface $statisticsProvider,
+        SalesDataProviderInterface $salesDataProvider,
+        MoneyFormatterInterface $moneyFormatter
+    ) {
+        $this->statisticsProvider = $statisticsProvider;
+        $this->salesDataProvider = $salesDataProvider;
+        $this->moneyFormatter = $moneyFormatter;
+    }
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:33]:
```suggestion
```

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 23]
```diff
@@ -0,0 +1,66 @@
...
+    public function getRawData(ChannelInterface $channel, \DateTime $startDate, \DateTime $endDate, string $interval): array
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:34]:
```suggestion
    public function getRawData(ChannelInterface $channel, \DateTimeInterface $startDate, \DateTimeInterface $endDate, string $interval): array
```

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 24]
```diff
@@ -0,0 +1,82 @@
...
+use Sylius\Bundle\MoneyBundle\Formatter\MoneyFormatterInterface;
+use Sylius\Component\Core\Dashboard\DashboardStatistics;
+use Sylius\Component\Core\Dashboard\DashboardStatisticsProviderInterface;
+use Sylius\Component\Core\Dashboard\Interval;
+use Sylius\Component\Core\Dashboard\SalesDataProviderInterface;
+use Sylius\Component\Core\Dashboard\SalesSummaryInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Currency\Model\CurrencyInterface;
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:16]:
```suggestion
```

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 25]
```diff
@@ -0,0 +1,82 @@
+<?php
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:2]: Missing license docblock

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 26]
```diff
@@ -0,0 +1,82 @@
...
+
+class StatisticsDataProviderSpec extends ObjectBehavior
+{
+    function let(DashboardStatisticsProviderInterface $statisticsProvider, SalesDataProviderInterface $salesDataProvider, MoneyFormatterInterface $moneyFormatter): void
+    {
+        $this->beConstructedWith($statisticsProvider, $salesDataProvider, $moneyFormatter);
+    }
+
+    function it_is_initializable()
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:24]: Missing return type

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 27]
```diff
@@ -0,0 +1,82 @@
...
+        $statistics->getNumberOfNewOrders()->willReturn(2);
+        $statistics->getNumberOfNewCustomers()->willReturn(21);
+
+        $moneyFormatter->format($totalSales, $currencyCode)->willReturn('$372.74');
+
+        $moneyFormatter->format($averageOrderValue, $currencyCode)->willReturn('$186.37');
+
+        $channel->getCode()->willReturn('FASHION_WEB');
+
+
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:64]:
```suggestion
```

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php (hunk 16 above)
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:43]: I would be for not using variables for strings and numbers

File: src/Sylius/Component/Core/Dashboard/DashboardStatistics.php [hunk 28]
```diff
@@ -24,14 +27,23 @@ class DashboardStatistics
...
     private $numberOfNewCustomers;
 
+    /** @var ChannelInterface */
+    private $channel;
+
     /**
      * @throws \InvalidArgumentException
      */
-    public function __construct(int $totalSales, int $numberOfNewOrders, int $numberOfNewCustomers)
+    public function __construct(int $totalSales, int $numberOfNewOrders, int $numberOfNewCustomers, ChannelInterface $channel = null)
```
Comment (GSadee) [inline src/Sylius/Component/Core/Dashboard/DashboardStatistics.php:36]:
```suggestion
    public function __construct(int $totalSales, int $numberOfNewOrders, int $numberOfNewCustomers, ?ChannelInterface $channel = null)
```

File: src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php (hunk 19 above)
Comment (SirDomin) [reply src/Sylius/Bundle/AdminBundle/Controller/Dashboard/StatisticsController.php]: nah we don't, removed them

File: src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_chart.html.twig [hunk 29]
```diff
@@ -1,12 +0,0 @@
-{% set labels, values = sales_summary.intervals, sales_summary.sales %}
```
Comment (GSadee) [inline src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_chart.html.twig:1]: Removing a template is a BC Break (https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html#templates)
You should add sth like at the begging of this template:
`{% deprecated "This template is deprecated since Sylius 1.8" %}`
and do not remove the file

Comment (SirDomin) [reply src/Sylius/Bundle/AdminBundle/Resources/views/Dashboard/_chart.html.twig:1]: got it back

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 30]
```diff
@@ -41,69 +39,62 @@ final class DashboardController
     /** @var SalesDataProviderInterface|null */
     private $salesDataProvider;
 
+    /** @var StatisticsDataProviderInterface */
+    private $statisticsDataProvider;
+
     public function __construct(
-        DashboardStatisticsProviderInterface $statisticsProvider,
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:45]: This is BC break - we cannot merge PR with this. Needs to be reverted

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 31]
```diff
@@ -41,69 +39,62 @@ final class DashboardController
...
+        ?SalesDataProviderInterface $salesDataProvider = null,
+        ?StatisticsDataProviderInterface $statisticsDataProvider = null
     ) {
-        $this->statisticsProvider = $statisticsProvider;
         $this->channelRepository = $channelRepository;
         $this->templatingEngine = $templatingEngine;
         $this->router = $router;
         $this->salesDataProvider = $salesDataProvider;
-
-        if ($this->salesDataProvider === null) {
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:57]: This note should be brought back IMHO. Why did you remove it? Similar note should be added for `statisticsDataProvider`

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php [hunk 32]
```diff
@@ -41,69 +39,62 @@ final class DashboardController
...
-                ->getSalesSummary($channel, $startDate, $endDate, $interval)
-            ;
-            $data['currency'] = $channel->getBaseCurrency()->getCode();
+
+        return $this->templatingEngine->renderResponse('@SyliusAdmin/Dashboard/index.html.twig', [
+            'channel' => $channel,
+        ]);
+    }
+
+    public function getRawData(Request $request): Response
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:73]:
```suggestion
    public function getRawDataAction(Request $request): Response
```

We tend to add `action` at the end of functions in controller. But this is a minor thing.

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 33]
```diff
@@ -0,0 +1,74 @@
...
+namespace Sylius\Bundle\AdminBundle\Provider;
+
+use Sylius\Bundle\MoneyBundle\Formatter\MoneyFormatterInterface;
+use Sylius\Component\Core\Dashboard\DashboardStatistics;
+use Sylius\Component\Core\Dashboard\DashboardStatisticsProviderInterface;
+use Sylius\Component\Core\Dashboard\Interval;
+use Sylius\Component\Core\Dashboard\SalesDataProviderInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+
+class StatisticsDataProvider implements StatisticsDataProviderInterface
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:21]:
```suggestion
final class StatisticsDataProvider implements StatisticsDataProviderInterface
```

File: src/Sylius/Bundle/AdminBundle/Resources/config/app/config.yml [hunk 34]
```diff
@@ -906,28 +906,25 @@ sylius_ui:
             blocks:
                 before_header_legacy:
                     template: "@SyliusUi/Block/_legacySonataEvent.html.twig"
-                    priority: 55
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/Resources/config/app/config.yml:909]: I'm not sure how if we should change already defined priorities. We may mess up someones UI

File: src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php [hunk 35]
```diff
@@ -0,0 +1,243 @@
...
+use Sylius\Bundle\MoneyBundle\Formatter\MoneyFormatterInterface;
+use Sylius\Component\Core\Dashboard\DashboardStatistics;
+use Sylius\Component\Core\Dashboard\DashboardStatisticsProviderInterface;
+use Sylius\Component\Core\Dashboard\Interval;
+use Sylius\Component\Core\Dashboard\SalesDataProviderInterface;
+use Sylius\Component\Core\Dashboard\SalesSummaryInterface;
+use Sylius\Component\Core\Model\ChannelInterface;
+use Sylius\Component\Currency\Model\CurrencyInterface;
+
+class StatisticsDataProviderSpec extends ObjectBehavior
```
Comment (lchrusciel) [inline src/Sylius/Bundle/AdminBundle/spec/Provider/StatisticsDataProviderSpec.php:26]:
```suggestion
final class StatisticsDataProviderSpec extends ObjectBehavior
```

File: src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php (hunk 30 above)
Comment (GSadee) [reply src/Sylius/Bundle/AdminBundle/Controller/DashboardController.php:45]: According to our [BC Promise](https://docs.sylius.com/en/latest/book/organization/backward-compatibility-promise.html#final-controllers) it is not a BC break if I'm not wrong

**Comment by lchrusciel** [comment]
Thanks, @SirDomin! :1st_place_medal:

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php [hunk 36]
```diff
@@ -0,0 +1,74 @@
...
+        $currencyCode = $channel->getBaseCurrency()->getCode();
+
+        return [
+            'sales_summary' => [
+                'intervals' => $salesSummary->getIntervals(),
+                'sales' => $salesSummary->getSales()
+            ],
+            'channel' => [
+                'base_currency_code' => $currencyCode,
+                'channel_code' => $channel->getCode(),
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProvider.php:64]: `channel_code` -> `code`?

File: src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProviderInterface.php [hunk 37]
```diff
@@ -0,0 +1,12 @@
...
+
+declare(strict_types=1);
+
+namespace Sylius\Bundle\AdminBundle\Provider;
+
+use Sylius\Component\Core\Model\ChannelInterface;
+
+interface StatisticsDataProviderInterface
+{
+    public function getRawData(ChannelInterface $channel, \DateTimeInterface $startDate, \DateTimeInterface $endDate, string $interval): array;
```
Comment (pamil) [inline src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProviderInterface.php:11]: `array` return type on an interface makes it almost impossible to maintain BC, we should define its structure via a docblock or VO.

Comment (lchrusciel) [reply src/Sylius/Bundle/AdminBundle/Provider/StatisticsDataProviderInterface.php:11]: I believe, the problem occurs, because we are doing a mapping from VO to array due to the serialization problem. Removing this constraint and usage of JMS/Symfony Serializer should resolve the problem
